src/main/java/io/github/opencubicchunks/cubicchunks/world/level/chunk/ProtoCube.java

```diff
@@ -238,23 +241,48 @@ public void onEnteringFeaturesStatus() {
     }
 
     @Override
-    public void sectionLoaded(SurfaceTrackerSection surfaceTrackerSection, int localSectionX, int localSectionZ) {
+    public void sectionLoaded(@Nonnull SurfaceTrackerLeaf surfaceTrackerLeaf, int localSectionX, int localSectionZ) {
         int idx = localSectionX + localSectionZ * DIAMETER_IN_SECTIONS;
 
-        if (surfaceTrackerSection.getRawType() == -1) { //light
-            this.lightHeightmaps[idx] = (LightSurfaceTrackerSection) surfaceTrackerSection;
+        if (surfaceTrackerLeaf.getRawType() == -1) { //light
+            this.lightHeightmaps[idx] = surfaceTrackerLeaf;
         } else { // normal heightmap
-            this.heightmaps.computeIfAbsent(surfaceTrackerSection.getType(),
-                type -> new SurfaceTrackerSection[DIAMETER_IN_SECTIONS * DIAMETER_IN_SECTIONS]
-            )[idx] = surfaceTrackerSection;
+            this.heightmaps.computeIfAbsent(surfaceTrackerLeaf.getType(),
+                type -> new SurfaceTrackerLeaf[DIAMETER_IN_SECTIONS * DIAMETER_IN_SECTIONS]
+            )[idx] = surfaceTrackerLeaf;
         }
     }
 
-    public LightSurfaceTrackerSection[] getLightHeightmaps() {
+    @Override
+    public void unloadNode(@Nonnull HeightmapStorage storage) {
+        for (SurfaceTrackerLeaf[] heightmapLeaves : this.heightmaps.values()) {
+            for (int localSectionZ = 0; localSectionZ < CubeAccess.DIAMETER_IN_SECTIONS; localSectionZ++) {
+                for (int localSectionX = 0; localSectionX < CubeAccess.DIAMETER_IN_SECTIONS; localSectionX++) {
+                    int i = localSectionX + localSectionZ * CubeAccess.DIAMETER_IN_SECTIONS;
+                    if (heightmapLeaves[i] != null) {
+                        heightmapLeaves[i].cubeUnloaded(localSectionX, localSectionZ, storage);
+                        heightmapLeaves[i] = null;
+                    }
+                }
+            }
+        }
+        SurfaceTrackerLeaf[] lightHeightmapLeaves = this.lightHeightmaps;
+        for (int localSectionZ = 0; localSectionZ < CubeAccess.DIAMETER_IN_SECTIONS; localSectionZ++) {
+            for (int localSectionX = 0; localSectionX < CubeAccess.DIAMETER_IN_SECTIONS; localSectionX++) {
+                int i = localSectionX + localSectionZ * CubeAccess.DIAMETER_IN_SECTIONS;
+                if (lightHeightmapLeaves[i] != null) {
+                    lightHeightmapLeaves[i].cubeUnloaded(localSectionX, localSectionZ, storage);
+                    lightHeightmapLeaves[i] = null;
+                }
+            }
+        }
```

Comment:
to avoid duplication like this, may be better to have a method that returns an iterable of all heightmaps if this is done in  more places.

Reply:
I assume you're referring to the normal/light heightmaps? Yeah that could be nice, though we'd need a `LIGHT` heightmap type, or return a byte map or something

Reply:
Made a note about this in [the project](https://github.com/orgs/OpenCubicChunks/projects/1)